envoy/adapter/adapter.go

```diff
@@ -201,23 +199,89 @@ func envoyListenerFromService(svc *service.Service, envoyServiceName string,
 			},
 		}
 	case "tcp":
-		connectionManagerName = wellknown.TCPProxy
+		managerName = wellknown.TCPProxy
 
-		connectionManager = &tcpp.TcpProxy{
+		manager = &tcpp.TcpProxy{
 			StatPrefix: "ingress_tcp",
 			ClusterSpecifier: &tcpp.TcpProxy_Cluster{
 				Cluster: envoyServiceName,
 			},
 		}
+	case "ws":
+		managerName = wellknown.HTTPConnectionManager
+
+		manager = &hcm.HttpConnectionManager{
+			StatPrefix: "ingress_http",
+			HttpFilters: []*hcm.HttpFilter{{
+				Name: wellknown.Router,
+			}},
+			RouteSpecifier: &hcm.HttpConnectionManager_RouteConfig{
+				RouteConfig: &api.RouteConfiguration{
+					ValidateClusters: &wrappers.BoolValue{Value: false},
+					VirtualHosts: []*route.VirtualHost{{
+						Name:    svc.Name,
+						Domains: []string{"*"},
+						Routes: []*route.Route{{
+							Match: &route.RouteMatch{
+								PathSpecifier: &route.RouteMatch_Prefix{
+									Prefix: "/",
+								},
+							},
+							Action: &route.Route_Route{
+								Route: &route.RouteAction{
+									ClusterSpecifier: &route.RouteAction_Cluster{
+										Cluster: envoyServiceName,
+									},
+									Timeout: &duration.Duration{},
+								},
+							},
+						}},
+					}},
+				},
+			},
+			UpgradeConfigs: []*hcm.HttpConnectionManager_UpgradeConfig{
+				{
+					UpgradeType: "websocket",
+				},
```

Comment:
This the majority of the needed change!

Reply:
Probably not worth the effort, but, to avoid duplication, you could leave the switch statement as it was and add `case "ws"` on top as fallthrough to `case "http"` and then put an `if svc.ProxyMode == "ws"` statement inside `case "http"` to populate `manager. UpgradeConfigs` with the `HttpConnectionManager_UpgradeConfig`.